src/stream/src/executor/top_n/utils.rs

```diff
@@ -154,23 +156,44 @@ pub fn generate_output(
     }
 }
 
-pub fn generate_executor_pk_indices_info(
-    order_pairs: &[OrderPair],
-    schema: &Schema,
-) -> (PkIndices, Vec<DataType>, Vec<OrderType>) {
+pub fn generate_executor_pk_indices_info(order_pairs: &[OrderPair]) -> (PkIndices, Vec<OrderType>) {
     let mut internal_key_indices = vec![];
     let mut internal_order_types = vec![];
     for order_pair in order_pairs {
         internal_key_indices.push(order_pair.column_idx);
         internal_order_types.push(order_pair.order_type);
     }
-    let internal_data_types = internal_key_indices
-        .iter()
-        .map(|idx| schema.fields()[*idx].data_type())
-        .collect();
-    (
-        internal_key_indices,
-        internal_data_types,
-        internal_order_types,
-    )
+    (internal_key_indices, internal_order_types)
+}
+
+/// For a given pk (Row), it can be split into `order_key` and `additional_pk` according to
+/// `order_by_len`, and the two split parts are serialized separately.
+pub fn serialize_pk_to_cache_key(
+    pk: Row,
+    order_by_len: usize,
+    pk_date_types: &[DataType],
+    pk_order_types: &[OrderType],
+) -> CacheKey {
+    let (first_key_data_types, second_key_data_types) = pk_date_types.split_at(order_by_len);
+    let (first_key_order_types, second_key_order_types) = pk_order_types.split_at(order_by_len);
+    let first_key_serde = OrderedRowSerde::new(
+        first_key_data_types.to_vec(),
+        first_key_order_types.to_vec(),
+    );
+    let second_key_serde = OrderedRowSerde::new(
+        second_key_data_types.to_vec(),
+        second_key_order_types.to_vec(),
+    );
```

Comment:
I think we should construct the `OrderedRowSerde` when the state table constructs only one time instead of constructing for each row.